File: graphlib-3/go/lib/graph.go
package graphlib

import (
	"sort"
)

// CreateGraph creates a new graph
func CreateGraph(directed, weighted bool) map[string]interface{} {
	id := Store.GenerateID()
	g := &Graph{
		ID:       id,
		Directed: directed,
		Weighted: weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InAdj:    make(map[string]map[string]float64),
	}
	Store.AddGraph(g)
	return map[string]interface{}{
		"id":         id,
		"directed":   directed,
		"weighted":   weighted,
		"node_count": 0,
		"edge_count": 0,
	}
}

// AddNode adds a node to the graph
func AddNode(graphID, nodeID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}
	if g.Nodes[nodeID] {
		return map[string]interface{}{"success": false, "error": "node_already_exists"}
	}
	g.Nodes[nodeID] = true
	if g.Adj[nodeID] == nil {
		g.Adj[nodeID] = make(map[string]float64)
	}
	if g.InAdj[nodeID] == nil {
		g.InAdj[nodeID] = make(map[string]float64)
	}
	Store.Save()
	return map[string]interface{}{"success": true, "node_id": nodeID}
}

// AddEdge adds an edge to the graph
func AddEdge(graphID, from, to string, weight *float64) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}

	// Auto-create nodes if they don't exist
	if !g.Nodes[from] {
		g.Nodes[from] = true
		g.Adj[from] = make(map[string]float64)
		g.InAdj[from] = make(map[string]float64)
	}
	if !g.Nodes[to] {
		g.Nodes[to] = true
		g.Adj[to] = make(map[string]float64)
		g.InAdj[to] = make(map[string]float64)
	}

	// Check if edge already exists
	if _, exists := g.Adj[from][to]; exists {
		return map[string]interface{}{"success": false, "error": "edge_already_exists"}
	}

	w := 1.0
	if weight != nil {
		w = *weight
	}

	// Add edge
	g.Adj[from][to] = w
	g.InAdj[to][from] = w

	// For undirected, add reverse edge
	if !g.Directed {
		g.Adj[to][from] = w
		g.InAdj[from][to] = w
	}

	Store.Save()
	return map[string]interface{}{"success": true, "from": from, "to": to, "weight": w}
}

// RemoveNode removes a node and all its incident edges
func RemoveNode(graphID, nodeID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}
	if !g.Nodes[nodeID] {
		return map[string]interface{}{"success": false, "error": "node_not_found"}
	}

	removedEdges := 0

	// For undirected graphs, we need to count edges once (not twice)
	// and remove from both Adj and InAdj of neighbors
	if g.Directed {
		// Directed: count outgoing edges
		for neighbor := range g.Adj[nodeID] {
			delete(g.InAdj[neighbor], nodeID)
			removedEdges++
		}
		// Directed: count incoming edges
		for neighbor := range g.InAdj[nodeID] {
			if neighbor != nodeID { // Don't double count self-loop
				delete(g.Adj[neighbor], nodeID)
				removedEdges++
			}
		}
	} else {
		// Undirected: edges are stored in both directions, count once
		for neighbor := range g.Adj[nodeID] {
			delete(g.InAdj[neighbor], nodeID)
			delete(g.Adj[neighbor], nodeID)
			removedEdges++
		}
	}

	delete(g.Nodes, nodeID)
	delete(g.Adj, nodeID)
	delete(g.InAdj, nodeID)

	Store.Save()
	return map[string]interface{}{"success": true, "removed_edges": removedEdges}
}

// RemoveEdge removes an edge from the graph
func RemoveEdge(graphID, from, to string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}

	if _, exists := g.Adj[from][to]; !exists {
		return map[string]interface{}{"success": false, "error": "edge_not_found"}
	}

	delete(g.Adj[from], to)
	delete(g.InAdj[to], from)

	if !g.Directed {
		delete(g.Adj[to], from)
		delete(g.InAdj[from], to)
	}

	Store.Save()
	return map[string]interface{}{"success": true}
}

// GetNodes returns all nodes in the graph
func GetNodes(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"nodes": []string{}, "count": 0}
	}

	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	return map[string]interface{}{"nodes": nodes, "count": len(nodes)}
}

// GetEdges returns all edges in the graph
func GetEdges(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"edges": []Edge{}, "count": 0}
	}

	edges := []Edge{}
	seen := make(map[string]bool)

	// Get sorted nodes for deterministic order
	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	for _, from := range nodes {
		// Get sorted neighbors
		neighbors := make([]string, 0, len(g.Adj[from]))
		for to := range g.Adj[from] {
			neighbors = append(neighbors, to)
		}
		sort.Strings(neighbors)

		for _, to := range neighbors {
			weight := g.Adj[from][to]
			key := from + "->" + to
			reverseKey := to + "->" + from

			if !g.Directed {
				// For undirected, only add edge once
				if !seen[key] && !seen[reverseKey] {
					edges = append(edges, Edge{From: from, To: to, Weight: weight})
					seen[key] = true
				}
			} else {
				edges = append(edges, Edge{From: from, To: to, Weight: weight})
			}
		}
	}

	return map[string]interface{}{"edges": edges, "count": len(edges)}
}

// GetNeighbors returns neighbors of a node
func GetNeighbors(graphID, nodeID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"neighbors": []string{}, "count": 0}
	}

	neighbors := make([]string, 0)
	for neighbor := range g.Adj[nodeID] {
		neighbors = append(neighbors, neighbor)
	}
	sort.Strings(neighbors)

	return map[string]interface{}{"neighbors": neighbors, "count": len(neighbors)}
}

// HasNode checks if a node exists
func HasNode(graphID, nodeID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"exists": false}
	}
	return map[string]interface{}{"exists": g.Nodes[nodeID]}
}

// HasEdge checks if an edge exists
func HasEdge(graphID, from, to string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"exists": false}
	}
	weight, exists := g.Adj[from][to]
	if !exists {
		return map[string]interface{}{"exists": false}
	}
	return map[string]interface{}{"exists": true, "weight": weight}
}

// GetDegree returns the degree of a node
func GetDegree(graphID, nodeID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"degree": 0, "in_degree": 0, "out_degree": 0}
	}

	outDegree := len(g.Adj[nodeID])
	inDegree := len(g.InAdj[nodeID])

	if !g.Directed {
		// For undirected graphs, in_degree == out_degree == degree
		return map[string]interface{}{
			"degree":     outDegree,
			"in_degree":  outDegree,
			"out_degree": outDegree,
		}
	}

	return map[string]interface{}{
		"degree":     inDegree + outDegree,
		"in_degree":  inDegree,
		"out_degree": outDegree,
	}
}

// GetGraphInfo returns information about a graph
func GetGraphInfo(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"error": "graph_not_found"}
	}

	edgeCount := 0
	for node := range g.Adj {
		edgeCount += len(g.Adj[node])
	}
	if !g.Directed {
		edgeCount /= 2
	}

	isConnectedResult := IsConnected(graphID)
	hasCycleResult := HasCycle(graphID)

	return map[string]interface{}{
		"id":           g.ID,
		"directed":     g.Directed,
		"weighted":     g.Weighted,
		"node_count":   len(g.Nodes),
		"edge_count":   edgeCount,
		"is_connected": isConnectedResult["is_connected"],
		"has_cycle":    hasCycleResult["has_cycle"],
	}
}

// ClearGraph removes all nodes and edges
func ClearGraph(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}

	g.Nodes = make(map[string]bool)
	g.Adj = make(map[string]map[string]float64)
	g.InAdj = make(map[string]map[string]float64)

	Store.Save()
	return map[string]interface{}{"success": true}
}

// CloneGraph creates a copy of a graph
func CloneGraph(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"error": "graph_not_found"}
	}

	newID := Store.GenerateID()
	newGraph := &Graph{
		ID:       newID,
		Directed: g.Directed,
		Weighted: g.Weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InAdj:    make(map[string]map[string]float64),
	}

	// Copy nodes
	for node := range g.Nodes {
		newGraph.Nodes[node] = true
		newGraph.Adj[node] = make(map[string]float64)
		newGraph.InAdj[node] = make(map[string]float64)
	}

	// Copy edges
	for from, neighbors := range g.Adj {
		for to, weight := range neighbors {
			newGraph.Adj[from][to] = weight
		}
	}
	for to, neighbors := range g.InAdj {
		for from, weight := range neighbors {
			newGraph.InAdj[to][from] = weight
		}
	}

	Store.AddGraph(newGraph)

	edgeCount := 0
	for node := range newGraph.Adj {
		edgeCount += len(newGraph.Adj[node])
	}
	if !newGraph.Directed {
		edgeCount /= 2
	}

	return map[string]interface{}{
		"id":         newID,
		"directed":   newGraph.Directed,
		"weighted":   newGraph.Weighted,
		"node_count": len(newGraph.Nodes),
		"edge_count": edgeCount,
	}
}

// Subgraph creates a subgraph containing only specified nodes
func Subgraph(graphID string, nodes []string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"error": "graph_not_found"}
	}

	newID := Store.GenerateID()
	newGraph := &Graph{
		ID:       newID,
		Directed: g.Directed,
		Weighted: g.Weighted,
		Nodes:    make(map[string]bool),
		Adj:      make(map[string]map[string]float64),
		InAdj:    make(map[string]map[string]float64),
	}

	// Create node set for quick lookup
	nodeSet := make(map[string]bool)
	for _, node := range nodes {
		if g.Nodes[node] {
			nodeSet[node] = true
			newGraph.Nodes[node] = true
			newGraph.Adj[node] = make(map[string]float64)
			newGraph.InAdj[node] = make(map[string]float64)
		}
	}

	// Copy edges between nodes in the subgraph
	for from := range nodeSet {
		for to, weight := range g.Adj[from] {
			if nodeSet[to] {
				newGraph.Adj[from][to] = weight
				newGraph.InAdj[to][from] = weight
			}
		}
	}

	Store.AddGraph(newGraph)

	edgeCount := 0
	for node := range newGraph.Adj {
		edgeCount += len(newGraph.Adj[node])
	}
	if !newGraph.Directed {
		edgeCount /= 2
	}

	return map[string]interface{}{
		"id":         newID,
		"directed":   newGraph.Directed,
		"weighted":   newGraph.Weighted,
		"node_count": len(newGraph.Nodes),
		"edge_count": edgeCount,
	}
}

File: graphlib-3/ts/src/graph.ts
import * as fs from 'fs';
import * as path from 'path';
import {
  Graph,
  GraphOptions,
  CreateGraphResult,
  AddNodeResult,
  AddEdgeResult,
  RemoveNodeResult,
  RemoveEdgeResult,
  GetNodesResult,
  GetEdgesResult,
  GetNeighborsResult,
  HasNodeResult,
  HasEdgeResult,
  GetDegreeResult,
  ClearGraphResult,
  CloneGraphResult,
  SubgraphResult,
  GraphInfoResult,
  Edge,
} from './types';

const DATA_FILE = path.join(__dirname, '..', 'data', 'graphs.json');

interface SerializedGraph {
  id: string;
  directed: boolean;
  weighted: boolean;
  nodes: string[];
  edges: Array<{ from: string; to: string; weight: number }>;
}

interface DataStore {
  graphs: Record<string, SerializedGraph>;
  counter: number;
}

function ensureDataDir(): void {
  const dir = path.dirname(DATA_FILE);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

function loadData(): DataStore {
  ensureDataDir();
  if (fs.existsSync(DATA_FILE)) {
    const content = fs.readFileSync(DATA_FILE, 'utf8');
    return JSON.parse(content);
  }
  return { graphs: {}, counter: 0 };
}

function saveData(data: DataStore): void {
  ensureDataDir();
  fs.writeFileSync(DATA_FILE, JSON.stringify(data), 'utf8');
}

function serializeGraph(g: Graph): SerializedGraph {
  const edges: Array<{ from: string; to: string; weight: number }> = [];
  const seen = new Set<string>();
  for (const [from, neighbors] of g.adjacency) {
    for (const [to, weight] of neighbors) {
      const key = g.directed ? `${from}->${to}` : [from, to].sort().join('-');
      if (!seen.has(key)) {
        seen.add(key);
        edges.push({ from, to, weight });
      }
    }
  }
  return {
    id: g.id,
    directed: g.directed,
    weighted: g.weighted,
    nodes: Array.from(g.nodes),
    edges,
  };
}

function deserializeGraph(s: SerializedGraph): Graph {
  const g: Graph = {
    id: s.id,
    directed: s.directed,
    weighted: s.weighted,
    nodes: new Set(s.nodes),
    adjacency: new Map(),
    inEdges: new Map(),
  };
  for (const node of s.nodes) {
    g.adjacency.set(node, new Map());
    g.inEdges.set(node, new Map());
  }
  for (const e of s.edges) {
    g.adjacency.get(e.from)!.set(e.to, e.weight);
    g.inEdges.get(e.to)!.set(e.from, e.weight);
    if (!s.directed && e.from !== e.to) {
      g.adjacency.get(e.to)!.set(e.from, e.weight);
      g.inEdges.get(e.from)!.set(e.to, e.weight);
    }
  }
  return g;
}

export function getGraph(graphId: string): Graph | undefined {
  const data = loadData();
  const s = data.graphs[graphId];
  if (!s) return undefined;
  return deserializeGraph(s);
}

function saveGraph(g: Graph): void {
  const data = loadData();
  data.graphs[g.id] = serializeGraph(g);
  saveData(data);
}

export function createGraph(options: GraphOptions = {}): CreateGraphResult {
  const data = loadData();
  data.counter++;
  const id = `graph-${data.counter}`;
  const graph: SerializedGraph = {
    id,
    directed: options.directed ?? false,
    weighted: options.weighted ?? false,
    nodes: [],
    edges: [],
  };
  data.graphs[id] = graph;
  saveData(data);
  return {
    id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: 0,
    edge_count: 0,
  };
}

export function addNode(graphId: string, nodeId: string): AddNodeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }
  if (graph.nodes.has(nodeId)) {
    return { success: false, error: 'node_already_exists' };
  }
  graph.nodes.add(nodeId);
  graph.adjacency.set(nodeId, new Map());
  graph.inEdges.set(nodeId, new Map());
  saveGraph(graph);
  return { success: true, node_id: nodeId };
}

export function addEdge(
  graphId: string,
  fromNode: string,
  toNode: string,
  weight?: number
): AddEdgeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  // Auto-create nodes
  if (!graph.nodes.has(fromNode)) {
    graph.nodes.add(fromNode);
    graph.adjacency.set(fromNode, new Map());
    graph.inEdges.set(fromNode, new Map());
  }
  if (!graph.nodes.has(toNode)) {
    graph.nodes.add(toNode);
    graph.adjacency.set(toNode, new Map());
    graph.inEdges.set(toNode, new Map());
  }

  const edgeWeight = weight ?? 1.0;
  const fromAdj = graph.adjacency.get(fromNode)!;

  // Check for existing edge
  if (fromAdj.has(toNode)) {
    return { success: false, error: 'edge_already_exists' };
  }

  // Add edge
  fromAdj.set(toNode, edgeWeight);
  graph.inEdges.get(toNode)!.set(fromNode, edgeWeight);

  // For undirected, add reverse edge
  if (!graph.directed && fromNode !== toNode) {
    graph.adjacency.get(toNode)!.set(fromNode, edgeWeight);
    graph.inEdges.get(fromNode)!.set(toNode, edgeWeight);
  }

  saveGraph(graph);
  return { success: true, from: fromNode, to: toNode, weight: edgeWeight };
}

export function removeNode(graphId: string, nodeId: string): RemoveNodeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }
  if (!graph.nodes.has(nodeId)) {
    return { success: false, error: 'node_not_found' };
  }

  let removedEdges = 0;

  // Count and remove outgoing edges
  const outgoing = graph.adjacency.get(nodeId);
  if (outgoing) {
    removedEdges += outgoing.size;
    for (const neighbor of outgoing.keys()) {
      graph.inEdges.get(neighbor)?.delete(nodeId);
    }
  }

  // Count and remove incoming edges
  const incoming = graph.inEdges.get(nodeId);
  if (incoming) {
    for (const neighbor of incoming.keys()) {
      if (graph.directed || neighbor === nodeId) {
        removedEdges++;
      }
      graph.adjacency.get(neighbor)?.delete(nodeId);
    }
  }

  graph.nodes.delete(nodeId);
  graph.adjacency.delete(nodeId);
  graph.inEdges.delete(nodeId);

  saveGraph(graph);
  return { success: true, removed_edges: removedEdges };
}

export function removeEdge(
  graphId: string,
  fromNode: string,
  toNode: string
): RemoveEdgeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  const fromAdj = graph.adjacency.get(fromNode);
  if (!fromAdj || !fromAdj.has(toNode)) {
    return { success: false, error: 'edge_not_found' };
  }

  fromAdj.delete(toNode);
  graph.inEdges.get(toNode)?.delete(fromNode);

  if (!graph.directed && fromNode !== toNode) {
    graph.adjacency.get(toNode)?.delete(fromNode);
    graph.inEdges.get(fromNode)?.delete(toNode);
  }

  saveGraph(graph);
  return { success: true };
}

export function getNodes(graphId: string): GetNodesResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { nodes: [], count: 0 };
  }
  const nodes = Array.from(graph.nodes).sort();
  return { nodes, count: nodes.length };
}

export function getEdges(graphId: string): GetEdgesResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { edges: [], count: 0 };
  }

  const edges: Edge[] = [];
  const seen = new Set<string>();

  for (const [from, neighbors] of graph.adjacency) {
    for (const [to, weight] of neighbors) {
      const key = graph.directed ? `${from}->${to}` : [from, to].sort().join('-');
      if (!seen.has(key)) {
        seen.add(key);
        edges.push({ from, to, weight });
      }
    }
  }

  return { edges, count: edges.length };
}

export function getNeighbors(graphId: string, nodeId: string): GetNeighborsResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { neighbors: [], count: 0 };
  }
  const adj = graph.adjacency.get(nodeId);
  if (!adj) {
    return { neighbors: [], count: 0 };
  }
  const neighbors = Array.from(adj.keys()).sort();
  return { neighbors, count: neighbors.length };
}

export function hasNode(graphId: string, nodeId: string): HasNodeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false };
  }
  return { exists: graph.nodes.has(nodeId) };
}

export function hasEdge(
  graphId: string,
  fromNode: string,
  toNode: string
): HasEdgeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false };
  }
  const fromAdj = graph.adjacency.get(fromNode);
  if (!fromAdj || !fromAdj.has(toNode)) {
    return { exists: false };
  }
  return { exists: true, weight: fromAdj.get(toNode) };
}

export function getDegree(graphId: string, nodeId: string): GetDegreeResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { degree: 0, in_degree: 0, out_degree: 0 };
  }

  const outAdj = graph.adjacency.get(nodeId);
  const inAdj = graph.inEdges.get(nodeId);

  const outDegree = outAdj ? outAdj.size : 0;
  const inDegree = inAdj ? inAdj.size : 0;

  if (graph.directed) {
    return {
      degree: inDegree + outDegree,
      in_degree: inDegree,
      out_degree: outDegree,
    };
  } else {
    return {
      degree: outDegree,
      in_degree: outDegree,
      out_degree: outDegree,
    };
  }
}

export function getGraphInfo(graphId: string): GraphInfoResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return {
      id: graphId,
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
      is_connected: true,
      has_cycle: false,
    };
  }

  // Import these dynamically to avoid circular dependency
  const { hasCycle, isConnected } = require('./components');

  const edges = getEdges(graphId);
  const connResult = isConnected(graphId);
  const cycleResult = hasCycle(graphId);

  return {
    id: graph.id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: graph.nodes.size,
    edge_count: edges.count,
    is_connected: connResult.is_connected,
    has_cycle: cycleResult.has_cycle,
  };
}

export function clearGraph(graphId: string): ClearGraphResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false };
  }
  graph.nodes.clear();
  graph.adjacency.clear();
  graph.inEdges.clear();
  saveGraph(graph);
  return { success: true };
}

export function cloneGraph(graphId: string): CloneGraphResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return {
      id: '',
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
    };
  }

  const newGraphResult = createGraph({
    directed: graph.directed,
    weighted: graph.weighted,
  });

  const newGraph = getGraph(newGraphResult.id)!;

  // Copy nodes
  for (const node of graph.nodes) {
    newGraph.nodes.add(node);
    newGraph.adjacency.set(node, new Map());
    newGraph.inEdges.set(node, new Map());
  }

  // Copy edges
  const seen = new Set<string>();
  for (const [from, neighbors] of graph.adjacency) {
    for (const [to, weight] of neighbors) {
      const key = graph.directed ? `${from}->${to}` : [from, to].sort().join('-');
      if (!seen.has(key)) {
        seen.add(key);
        newGraph.adjacency.get(from)!.set(to, weight);
        newGraph.inEdges.get(to)!.set(from, weight);
        if (!graph.directed && from !== to) {
          newGraph.adjacency.get(to)!.set(from, weight);
          newGraph.inEdges.get(from)!.set(to, weight);
        }
      }
    }
  }

  saveGraph(newGraph);

  const edges = getEdges(newGraphResult.id);
  return {
    id: newGraphResult.id,
    directed: graph.directed,
    weighted: graph.weighted,
    node_count: graph.nodes.size,
    edge_count: edges.count,
  };
}

export function subgraph(graphId: string, nodeList: string[]): SubgraphResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return {
      id: '',
      directed: false,
      weighted: false,
      node_count: 0,
      edge_count: 0,
    };
  }

  const nodeSet = new Set(nodeList);
  const newGraphResult = createGraph({
    directed: graph.directed,
    weighted: graph.weighted,
  });

  const newGraph = getGraph(newGraphResult.id)!;

  // Add nodes that exist in original graph
  for (const node of nodeList) {
    if (graph.nodes.has(node)) {
      newGraph.nodes.add(node);
      newGraph.adjacency.set(node, new Map());
      newGraph.inEdges.set(node, new Map());
    }
  }

  // Add edges between selected nodes
  const seen = new Set<string>();
  for (const [from, neighbors] of graph.adjacency) {
    if (!nodeSet.has(from)) continue;
    for (const [to, weight] of neighbors) {
      if (!nodeSet.has(to)) continue;
      const key = graph.directed ? `${from}->${to}` : [from, to].sort().join('-');
      if (!seen.has(key)) {
        seen.add(key);
        newGraph.adjacency.get(from)!.set(to, weight);
        newGraph.inEdges.get(to)!.set(from, weight);
        if (!graph.directed && from !== to) {
          newGraph.adjacency.get(to)!.set(from, weight);
          newGraph.inEdges.get(from)!.set(to, weight);
        }
      }
    }
  }

  saveGraph(newGraph);

  const edges = getEdges(newGraphResult.id);
  return {
    id: newGraphResult.id,
    directed: newGraph.directed,
    weighted: newGraph.weighted,
    node_count: newGraph.nodes.size,
    edge_count: edges.count,
  };
}
